### PR TITLE
added password hashing

### PR DESCRIPTION
**Pull request by yutsevich**

### CODE REVIEW

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 1]
```diff
@@ -35,6 +36,9 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
 
         if (password.equals(repeatedPassword)) {
             User user = new User(name, login, password);
+            byte[] salt = HashUtil.getSalt();
+            user.setPassword(HashUtil.hashPassword(password, salt));
+            user.setSalt(salt);
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: Let's implement this logic in service layer

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        return hashedPassword.toString();
+    }
+
+    private boolean isValid(String password, byte[] salt) {
+        User user = new User();
+        user.setPassword(hashPassword(password, salt));
+        user.setSalt(salt);
+
+        return hashPassword(password,salt).equals(user.getPassword());
+    }
```
Comment (andriylagun) [inline src/main/java/com/internet/shop/util/HashUtil.java]: why do you need this method if you don't use it? And do you sure that u need logic for hashing in check method?

File: src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java [hunk 3]
```diff
@@ -91,13 +92,14 @@ public User create(User user) {
...
-        String query = "UPDATE users SET user_name=?, login=?"
-                + "WHERE password=? & deleted=FALSE";
+        String query = "UPDATE users SET user_name = ?, login = ?, password = ?, salt = ?"
+                + "WHERE user_id = ? AND deleted = false";
         try (Connection connection = ConnectionUtil.getConnection();
                 PreparedStatement statement = connection.prepareStatement(query)) {
             statement.setString(1, user.getName());
             statement.setString(2, user.getLogin());
             statement.setString(3, user.getPassword());
+            statement.setBytes(4, user.getSalt());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java:102]: should we also set userId?

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -15,7 +16,8 @@
     @Override
     public User login(String login, String password) throws AuthenticationException {
         Optional<User> user = userService.findByLogin(login);
-        if (user.isEmpty() || !user.get().getPassword().equals(password)) {
+        if (user.isEmpty() || !HashUtil.hashPassword(password, user.get().getSalt())
+                        .equals(user.get().getPassword())) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java]:
```suggestion
        if (user.isEmpty() || !isPasswordValid(user.get(), password)) {
```
let's create a private method here with this logic

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(HASHING_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            System.out.println("Invalid hashing algorithm " + HASHING_ALGORITHM);
+            e.printStackTrace();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/util/HashUtil.java]: let's throw a RuntimeException instead

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java (hunk 3 above)
Comment (yutsevich) [reply src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java:102]: no, it should be same because method finds user by id

Comment (KseniiaMakarova) [reply src/main/java/com/internet/shop/dao/mysqlimpl/UserDaoJdbcImpl.java:102]: you have a query with 5 `?` pieces of data that should be set if we want the statement to work properly
lI can see that you set only the first four

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 6]
```diff
@@ -25,7 +25,7 @@ public static String hashPassword(String password, byte[] salt) {
             }
         } catch (NoSuchAlgorithmException e) {
             System.out.println("Invalid hashing algorithm " + HASHING_ALGORITHM);
-            e.printStackTrace();
+            throw new RuntimeException();
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/util/HashUtil.java]: let's put an informative message here and also pass the original exception (`e`)

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**